src/cancelq/cancellable.h

```diff
@@ -0,0 +1,78 @@
+#ifndef CANCELLABLE_H
+#define CANCELLABLE_H
+
+#include <stddef.h>
+#include <stdbool.h>
+
+#define NOT_CANCELLABLE ((cancellable_id_t)-1)
+#define SCHEDULE_FAIL ((cancellable_id_t)-2)
+
+// The event queue that holds all these events
+typedef struct event_queue_s event_queue_t;
+
+// ID used to cancel cancellable_t
+typedef int cancellable_id_t;
+
+// Return a new event queue that will schedule and deliver events with
+// the eq functions below.
+event_queue_t *eq_new(void);
+
+// Frees an empty event queue. Fails if the queue is not empty.
+bool eq_free(event_queue_t *queue);
+
+// Schedule a cancellable event to be delivered at the appointed
+// time. Returns an ID that can be used to cancel the event. The ID is
```

Comment:
Seconds from January 1st 1970, or now?  At time, or after elapsed time?

Reply:
Separately, this is why I'm in favor of using some kind of `time_t` and `time_delta_t`, because I don't know by looking at the interface which it is, or if it's in seconds or milliseconds, or what, and what if my system has finer time resolution than milliseconds, what do I do, etc.

Reply:
This data structure doesn't care. It's just a number. Since it's used at insert and remove, the units cancel out. This way, one heap could be used as a delay timer, while another could use a wall clock.

Reply:
It really does care, though.  Given the interface to `eq_next_event`, it makes it impossible to actually use time deltas.  Here's an example to illustrate the issue: say the system is using milliseconds; one call to this interface sets a delay of 500, and some time passes, say 100ms, and a second call to this interface also wants a delay of 450ms.  Naturally, the first event should fire before the second one, but there is no way with this interface to make that happen without constantly decrementing all the stored time deltas, in which case you have to know they're time deltas, and then `eq_next_event` shouldn't take a number, because the only number it would act on is 0.  Therefore, this interface can only use absolute time.  That leaves us with a cron-like system (which is okay), but without timers (which is not okay).  Not only do I think we need both, I really believe deeply that `time_t` and `time_delta_t` are both necessary for each respective case.

I also feel like we're going in circles on this point, but I've gotten the notion that neither of us is going to budge, so it's not clear to me how to make progress.

Reply:
I think you're imagining something totally different from what I made. There's no changing the number associated with an event once it's in the heap. Time is kept outside the heap. Since the thing I implemented here is not cron-style timers, it's provably true that this mechanism can work for relative timers.

Here's how it works: time is kept by the system using one of several available clocks. For relative timers, the clock chosen is the system's monotonic uptime. See `smear.c` and `smeartime.h`.

If you wanted an absolute time, you'd provide it instead and all your timers would be cron-style timers.

I think the problem here is you're assuming the heap cares what the semantics of the time it's storing are. It doesn't. As long as a particular message queue is only ever given one type of time for all of its functions, it will work. This leaves room for an absolute timer to be a separate instance of the message queue.

Reply:
I don't believe I am imagining something different.  You say it uses the system's monotonic uptime.  That means it is not relative.  I.e., one would never call `eq_schedule(q, e, 500);`, but instead always call `eq_schedule(q, e, get_now_ns() + 1000 * 1000 * 500);`.  I would call that absolute.

Reply:
For posterity: this was resolved in gitter.